src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/AbstractTypeImportCompletionService.CacheEntry.cs

```diff
@@ -0,0 +1,150 @@
+                string language,
+                ImmutableArray<TypeImportCompletionItemInfo> items)
+            {
+                Checksum = checksum;
+                Language = language;
+
+                ItemInfos = items;
+            }
+
+            public ImmutableArray<CompletionItem> GetItemsForContext(
+                string language,
+                string genericTypeSuffix,
+                bool isInternalsVisible,
+                bool isAttributeContext,
+                bool isCaseSensitive)
+            {
+                // We will need to adjust some items if the request is made in:
+                // 1. attribute context, then we will not show or complete with "Attribute" suffix.
+                // 2. a project with different langauge than when the cache entry was created,
+                //    then we will change the generic suffix accordingly.
+                // Otherwise, we can simply return cached items.
+                var isSameLanguage = Language == language;
+                if (isSameLanguage && !isAttributeContext)
+                {
+                    return ItemInfos.Where(info => info.IsPublic || isInternalsVisible).SelectAsArray(info => info.Item);
+                }
+
+                var builder = ArrayBuilder<CompletionItem>.GetInstance();
+                foreach (var info in ItemInfos)
+                {
+                    if (info.IsPublic || isInternalsVisible)
+                    {
+                        var item = info.Item;
+                        if (isAttributeContext)
+                        {
+                            if (!info.IsAttribute)
+                            {
+                                continue;
+                            }
+
+                            item = GetAppropriateAttributeItem(info.Item, isCaseSensitive);
+                        }
+
+                        if (!isSameLanguage && info.IsGeneric)
+                        {
+                            // We don't want to cache this item.
+                            item = ImportCompletionItem.CreateItemWithGenericDisplaySuffix(item, genericTypeSuffix);
+                        }
+
+                        builder.Add(item);
+                    }
+                }
+
+                return builder.ToImmutableAndFree();
+
+                static CompletionItem GetAppropriateAttributeItem(CompletionItem attributeItem, bool isCaseSensitive)
+                {
+                    if (attributeItem.DisplayText.TryGetWithoutAttributeSuffix(isCaseSensitive: isCaseSensitive, out var attributeNameWithoutSuffix))
+                    {
+                        // We don't want to cache this item.
+                        return ImportCompletionItem.CreateAttributeItemWithoutSuffix(attributeItem, attributeNameWithoutSuffix);
+                    }
+
+                    return attributeItem;
+                }
+            }
+
+            public class Builder : IDisposable
```

Comment:
can/should this be a struct?

Reply:
It can, but this is only required when we have a cache miss, which should be very rare (except for the first time invocation, but even in that case it's the same number of references, e.g. a few hundreds), so probably won't make much difference.

Reply:
sgtm.  #resolved